File: src/apt.rs
use std::collections::BTreeMap;
use std::io::Write;

use anyhow::{Context, bail};
use chrono::{DateTime, Utc};
use flate2::Compression;
use flate2::write::GzEncoder;
use md5::Md5;
use sha1::Sha1;
use sha2::{Digest, Sha256};

#[derive(Debug, Clone)]
pub struct RepositoryPath {
    pub distribution: String,
    pub component: String,
    pub architecture: String,
}

impl RepositoryPath {
    pub fn packages_path(&self) -> String {
        format!(
            "dists/{}/{}/binary-{}/Packages",
            self.distribution, self.component, self.architecture
        )
    }

    pub fn packages_gz_path(&self) -> String {
        format!("{}.gz", self.packages_path())
    }

    pub fn release_path(&self) -> String {
        format!("dists/{}/Release", self.distribution)
    }

    pub fn pool_path(&self, package: &str) -> String {
        let prefix = if package.starts_with("lib") && package.len() > 3 {
            &package[..4]
        } else {
            &package[..1]
        };
        format!("pool/{}/{}/{}", self.component, prefix, package)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PackageEntry {
    pub package: String,
    pub version: String,
    pub architecture: String,
    pub filename: String,
    pub size: u64,
    pub md5sum: String,
    pub sha1: String,
    pub sha256: String,
    pub maintainer: Option<String>,
    pub description: Option<String>,
    pub section: Option<String>,
    pub priority: Option<String>,
    pub installed_size: Option<u64>,
    pub depends: Option<String>,
    pub recommends: Option<String>,
    pub suggests: Option<String>,
    pub conflicts: Option<String>,
    pub breaks: Option<String>,
    pub replaces: Option<String>,
    pub provides: Option<String>,
    pub homepage: Option<String>,
    pub extra: BTreeMap<String, String>,
}

impl PackageEntry {
    pub fn from_fields(fields: &BTreeMap<String, String>) -> anyhow::Result<Self> {
        let package = required(fields, "Package")?.to_string();
        let version = required(fields, "Version")?.to_string();
        let architecture = required(fields, "Architecture")?.to_string();
        let filename = required(fields, "Filename")?.to_string();
        let size = required(fields, "Size")?
            .parse()
            .with_context(|| format!("invalid Size field for {package}"))?;
        let md5sum = required(fields, "MD5sum")?.to_string();
        let sha1 = required(fields, "SHA1")?.to_string();
        let sha256 = required(fields, "SHA256")?.to_string();

        let known = [
            "Package",
            "Version",
            "Architecture",
            "Filename",
            "Size",
            "MD5sum",
            "SHA1",
            "SHA256",
            "Maintainer",
            "Description",
            "Section",
            "Priority",
            "Installed-Size",
            "Depends",
            "Recommends",
            "Suggests",
            "Conflicts",
            "Breaks",
            "Replaces",
            "Provides",
            "Homepage",
        ];

        let extra = fields
            .iter()
            .filter(|(key, value)| !known.contains(&key.as_str()) && !value.is_empty())
            .map(|(key, value)| (key.clone(), value.clone()))
            .collect();

        Ok(Self {
            package,
            version,
            architecture,
            filename,
            size,
            md5sum,
            sha1,
            sha256,
            maintainer: fields.get("Maintainer").cloned(),
            description: fields.get("Description").cloned(),
            section: fields.get("Section").cloned(),
            priority: fields.get("Priority").cloned(),
            installed_size: fields
                .get("Installed-Size")
                .map(|value| value.parse())
                .transpose()
                .context("invalid Installed-Size field")?,
            depends: fields.get("Depends").cloned(),
            recommends: fields.get("Recommends").cloned(),
            suggests: fields.get("Suggests").cloned(),
            conflicts: fields.get("Conflicts").cloned(),
            breaks: fields.get("Breaks").cloned(),
            replaces: fields.get("Replaces").cloned(),
            provides: fields.get("Provides").cloned(),
            homepage: fields.get("Homepage").cloned(),
            extra,
        })
    }
}

#[derive(Debug, Clone)]
pub struct ReleaseFile {
    pub origin: String,
    pub label: String,
    pub description: String,
    pub distribution: String,
    pub components: Vec<String>,
    pub architectures: Vec<String>,
    pub date: DateTime<Utc>,
    pub files: Vec<ReleaseFileEntry>,
}

#[derive(Debug, Clone)]
pub struct ReleaseFileEntry {
    pub filename: String,
    pub size: usize,
    pub md5: String,
    pub sha1: String,
    pub sha256: String,
}

impl ReleaseFile {
    pub fn generate(&self) -> Vec<u8> {
        let mut out = String::new();
        out.push_str(&format!(
            "Origin: {}\n",
            defaulted(&self.origin, "APT S3 Repository")
        ));
        out.push_str(&format!(
            "Label: {}\n",
            defaulted(&self.label, "APT S3 Repository")
        ));
        out.push_str(&format!("Suite: {}\n", self.distribution));
        out.push_str(&format!("Codename: {}\n", self.distribution));
        out.push_str(&format!(
            "Date: {}\n",
            self.date.format("%a, %d %b %Y %H:%M:%S UTC")
        ));
        out.push_str(&format!(
            "Architectures: {}\n",
            self.architectures.join(" ")
        ));
        out.push_str(&format!("Components: {}\n", self.components.join(" ")));
        out.push_str(&format!(
            "Description: {}\n",
            defaulted(&self.description, "APT repository hosted on S3")
        ));

        if !self.files.is_empty() {
            out.push_str("MD5Sum:\n");
            for file in &self.files {
                out.push_str(&format!(
                    " {} {:16} {}\n",
                    file.md5, file.size, file.filename
                ));
            }
            out.push_str("SHA1:\n");
            for file in &self.files {
                out.push_str(&format!(
                    " {} {:16} {}\n",
                    file.sha1, file.size, file.filename
                ));
            }
            out.push_str("SHA256:\n");
            for file in &self.files {
                out.push_str(&format!(
                    " {} {:16} {}\n",
                    file.sha256, file.size, file.filename
                ));
            }
        }

        out.into_bytes()
    }
}

pub fn parse_control_paragraphs(content: &str) -> anyhow::Result<Vec<BTreeMap<String, String>>> {
    if content.trim().is_empty() {
        return Ok(Vec::new());
    }

    let normalized = content.replace("\r\n", "\n");
    normalized
        .split("\n\n")
        .filter(|p| !p.trim().is_empty())
        .map(parse_paragraph)
        .collect()
}

pub fn parse_packages(content: &str) -> anyhow::Result<Vec<PackageEntry>> {
    parse_control_paragraphs(content)?
        .iter()
        .map(PackageEntry::from_fields)
        .collect()
}

pub fn generate_packages_content(packages: &[PackageEntry]) -> String {
    if packages.is_empty() {
        return String::new();
    }

    let entries = packages
        .iter()
        .map(generate_package_entry)
        .collect::<Vec<_>>();
    format!("{}\n\n", entries.join("\n\n"))
}

pub fn compress_gzip(data: &[u8]) -> anyhow::Result<Vec<u8>> {
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(data)?;
    Ok(encoder.finish()?)
}

pub fn calculate_hashes(data: &[u8]) -> (String, String, String) {
    let md5 = hex::encode(Md5::digest(data));
    let sha1 = hex::encode(Sha1::digest(data));
    let sha256 = hex::encode(Sha256::digest(data));
    (md5, sha1, sha256)
}

fn parse_paragraph(paragraph: &str) -> anyhow::Result<BTreeMap<String, String>> {
    let mut fields = BTreeMap::<String, String>::new();
    let mut current_key: Option<String> = None;

    for raw_line in paragraph.lines() {
        if raw_line.trim().is_empty() {
            continue;
        }

        if raw_line.starts_with(' ') || raw_line.starts_with('\t') {
            let key = current_key
                .as_ref()
                .context("control continuation line without a preceding field")?;
            let value = fields.get_mut(key).expect("current key exists");
            value.push('\n');
            value.push_str(raw_line);
            continue;
        }

        let (key, value) = raw_line
            .split_once(':')
            .with_context(|| format!("invalid control field line: {raw_line}"))?;
        let key = key.trim();
        if key.is_empty() {
            bail!("empty control field name");
        }
        let key = key.to_string();
        fields.insert(key.clone(), value.trim_start().to_string());
        current_key = Some(key);
    }

    Ok(fields)
}

fn generate_package_entry(pkg: &PackageEntry) -> String {
    let mut lines = vec![
        format!("Package: {}", pkg.package),
        format!("Version: {}", pkg.version),
        format!("Architecture: {}", pkg.architecture),
    ];

    push_opt(&mut lines, "Maintainer", &pkg.maintainer);
    if let Some(installed_size) = pkg.installed_size {
        lines.push(format!("Installed-Size: {installed_size}"));
    }
    push_opt(&mut lines, "Depends", &pkg.depends);
    push_opt(&mut lines, "Recommends", &pkg.recommends);
    push_opt(&mut lines, "Suggests", &pkg.suggests);
    push_opt(&mut lines, "Conflicts", &pkg.conflicts);
    push_opt(&mut lines, "Breaks", &pkg.breaks);
    push_opt(&mut lines, "Replaces", &pkg.replaces);
    push_opt(&mut lines, "Provides", &pkg.provides);
    push_opt(&mut lines, "Section", &pkg.section);
    push_opt(&mut lines, "Priority", &pkg.priority);
    push_opt(&mut lines, "Homepage", &pkg.homepage);

    for (key, value) in &pkg.extra {
        lines.push(format!("{key}: {value}"));
    }

    lines.push(format!("Filename: {}", pkg.filename));
    lines.push(format!("Size: {}", pkg.size));
    lines.push(format!("MD5sum: {}", pkg.md5sum));
    lines.push(format!("SHA1: {}", pkg.sha1));
    lines.push(format!("SHA256: {}", pkg.sha256));
    push_opt(&mut lines, "Description", &pkg.description);

    lines.join("\n")
}

fn push_opt(lines: &mut Vec<String>, key: &str, value: &Option<String>) {
    if let Some(value) = value.as_ref().filter(|v| !v.is_empty()) {
        lines.push(format!("{key}: {value}"));
    }
}

fn required<'a>(fields: &'a BTreeMap<String, String>, key: &str) -> anyhow::Result<&'a str> {
    fields
        .get(key)
        .map(String::as_str)
        .filter(|v| !v.is_empty())
        .with_context(|| format!("package entry missing required '{key}' field"))
}

fn defaulted<'a>(value: &'a str, default: &'a str) -> &'a str {
    if value.is_empty() { default } else { value }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pool_path_uses_lib_prefix() {
        let repo = RepositoryPath {
            distribution: "stable".into(),
            component: "main".into(),
            architecture: "amd64".into(),
        };
        assert_eq!(repo.pool_path("nginx"), "pool/main/n/nginx");
        assert_eq!(repo.pool_path("libssl"), "pool/main/libs/libssl");
    }

    #[test]
    fn parses_and_generates_packages() {
        let content = "\
Package: pkg
Version: 1.0
Architecture: amd64
Filename: pool/main/p/pkg/pkg_1.0_amd64.deb
Size: 12
MD5sum: d41d8cd98f00b204e9800998ecf8427e
SHA1: da39a3ee5e6b4b0d3255bfef95601890afd80709
SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855

";
        let packages = parse_packages(content).unwrap();
        assert_eq!(packages.len(), 1);
        assert_eq!(packages[0].package, "pkg");
        assert!(generate_packages_content(&packages).contains("Package: pkg"));
    }
}

File: internal/apt/apt.go
package apt

import (
	"bytes"
	"compress/gzip"
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

type RepositoryPath struct {
	Distribution string
	Component    string
	Architecture string
}

func (r RepositoryPath) PackagesPath() string {
	return fmt.Sprintf("dists/%s/%s/binary-%s/Packages", r.Distribution, r.Component, r.Architecture)
}

func (r RepositoryPath) PackagesGzPath() string {
	return r.PackagesPath() + ".gz"
}

func (r RepositoryPath) ReleasePath() string {
	return fmt.Sprintf("dists/%s/Release", r.Distribution)
}

func (r RepositoryPath) PoolPath(pkg string) string {
	var prefix string
	if strings.HasPrefix(pkg, "lib") && len(pkg) > 3 {
		prefix = pkg[:4]
	} else if len(pkg) > 0 {
		prefix = pkg[:1]
	}
	return fmt.Sprintf("pool/%s/%s/%s", r.Component, prefix, pkg)
}

type PackageEntry struct {
	Package       string
	Version       string
	Architecture  string
	Filename      string
	Size          uint64
	MD5sum        string
	SHA1          string
	SHA256        string
	Maintainer    string
	Description   string
	Section       string
	Priority      string
	InstalledSize uint64
	Depends       string
	Recommends    string
	Suggests      string
	Conflicts     string
	Breaks        string
	Replaces      string
	Provides      string
	Homepage      string
	Extra         map[string]string
}

func PackageEntryFromFields(fields map[string]string) (*PackageEntry, error) {
	pkg, err := required(fields, "Package")
	if err != nil {
		return nil, err
	}
	version, err := required(fields, "Version")
	if err != nil {
		return nil, err
	}
	arch, err := required(fields, "Architecture")
	if err != nil {
		return nil, err
	}
	filename, err := required(fields, "Filename")
	if err != nil {
		return nil, err
	}
	sizeStr, err := required(fields, "Size")
	if err != nil {
		return nil, err
	}
	size, err := strconv.ParseUint(sizeStr, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid Size field for %s: %w", pkg, err)
	}
	md5sum, err := required(fields, "MD5sum")
	if err != nil {
		return nil, err
	}
	sha1Val, err := required(fields, "SHA1")
	if err != nil {
		return nil, err
	}
	sha256Val, err := required(fields, "SHA256")
	if err != nil {
		return nil, err
	}

	known := map[string]bool{
		"Package":        true,
		"Version":        true,
		"Architecture":   true,
		"Filename":       true,
		"Size":           true,
		"MD5sum":         true,
		"SHA1":           true,
		"SHA256":         true,
		"Maintainer":     true,
		"Description":    true,
		"Section":        true,
		"Priority":       true,
		"Installed-Size": true,
		"Depends":        true,
		"Recommends":     true,
		"Suggests":       true,
		"Conflicts":      true,
		"Breaks":         true,
		"Replaces":       true,
		"Provides":       true,
		"Homepage":       true,
	}

	var instSize uint64
	if isStr, ok := fields["Installed-Size"]; ok {
		instSize, _ = strconv.ParseUint(isStr, 10, 64)
	}

	extra := make(map[string]string)
	for k, v := range fields {
		if !known[k] && v != "" {
			extra[k] = v
		}
	}

	return &PackageEntry{
		Package:       pkg,
		Version:       version,
		Architecture:  arch,
		Filename:      filename,
		Size:          size,
		MD5sum:        md5sum,
		SHA1:          sha1Val,
		SHA256:        sha256Val,
		Maintainer:    fields["Maintainer"],
		Description:   fields["Description"],
		Section:       fields["Section"],
		Priority:      fields["Priority"],
		InstalledSize: instSize,
		Depends:       fields["Depends"],
		Recommends:    fields["Recommends"],
		Suggests:      fields["Suggests"],
		Conflicts:     fields["Conflicts"],
		Breaks:        fields["Breaks"],
		Replaces:      fields["Replaces"],
		Provides:      fields["Provides"],
		Homepage:      fields["Homepage"],
		Extra:         extra,
	}, nil
}

type ReleaseFileEntry struct {
	Filename string
	Size     int
	MD5      string
	SHA1     string
	SHA256   string
}

type ReleaseFile struct {
	Origin        string
	Label         string
	Description   string
	Distribution  string
	Components    []string
	Architectures []string
	Date          time.Time
	Files         []ReleaseFileEntry
}

func (r *ReleaseFile) Generate() []byte {
	var out strings.Builder
	fmt.Fprintf(&out, "Origin: %s\n", defaulted(r.Origin, "APT S3 Repository"))
	fmt.Fprintf(&out, "Label: %s\n", defaulted(r.Label, "APT S3 Repository"))
	fmt.Fprintf(&out, "Suite: %s\n", r.Distribution)
	fmt.Fprintf(&out, "Codename: %s\n", r.Distribution)
	fmt.Fprintf(&out, "Date: %s\n", r.Date.UTC().Format("Mon, 02 Jan 2006 15:04:05 UTC"))
	fmt.Fprintf(&out, "Architectures: %s\n", strings.Join(r.Architectures, " "))
	fmt.Fprintf(&out, "Components: %s\n", strings.Join(r.Components, " "))
	fmt.Fprintf(&out, "Description: %s\n", defaulted(r.Description, "APT repository hosted on S3"))

	if len(r.Files) > 0 {
		out.WriteString("MD5Sum:\n")
		for _, f := range r.Files {
			fmt.Fprintf(&out, " %s %16d %s\n", f.MD5, f.Size, f.Filename)
		}
		out.WriteString("SHA1:\n")
		for _, f := range r.Files {
			fmt.Fprintf(&out, " %s %16d %s\n", f.SHA1, f.Size, f.Filename)
		}
		out.WriteString("SHA256:\n")
		for _, f := range r.Files {
			fmt.Fprintf(&out, " %s %16d %s\n", f.SHA256, f.Size, f.Filename)
		}
	}

	return []byte(out.String())
}

func ParseControlParagraphs(content string) ([]map[string]string, error) {
	if strings.TrimSpace(content) == "" {
		return nil, nil
	}

	normalized := strings.ReplaceAll(content, "\r\n", "\n")
	paragraphs := strings.Split(normalized, "\n\n")

	var result []map[string]string
	for _, p := range paragraphs {
		if strings.TrimSpace(p) == "" {
			continue
		}
		fields, err := parseParagraph(p)
		if err != nil {
			return nil, err
		}
		result = append(result, fields)
	}
	return result, nil
}

func parseParagraph(paragraph string) (map[string]string, error) {
	fields := make(map[string]string)
	var currentKey string

	lines := strings.Split(paragraph, "\n")
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}

		if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
			if currentKey == "" {
				return nil, fmt.Errorf("control continuation line without a preceding field")
			}
			fields[currentKey] = fields[currentKey] + "\n" + line
			continue
		}

		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			return nil, fmt.Errorf("invalid control field line: %s", line)
		}

		key := strings.TrimSpace(parts[0])
		if key == "" {
			return nil, fmt.Errorf("empty control field name")
		}
		value := strings.TrimLeft(parts[1], " \t")
		fields[key] = value
		currentKey = key
	}

	return fields, nil
}

func ParsePackages(content string) ([]PackageEntry, error) {
	paras, err := ParseControlParagraphs(content)
	if err != nil {
		return nil, err
	}

	var entries []PackageEntry
	for _, fields := range paras {
		entry, err := PackageEntryFromFields(fields)
		if err != nil {
			return nil, err
		}
		entries = append(entries, *entry)
	}
	return entries, nil
}

func GeneratePackagesContent(packages []PackageEntry) string {
	if len(packages) == 0 {
		return ""
	}

	var entries []string
	for _, pkg := range packages {
		entries = append(entries, generatePackageEntry(&pkg))
	}
	return strings.Join(entries, "\n\n") + "\n\n"
}

func generatePackageEntry(pkg *PackageEntry) string {
	var lines []string
	lines = append(lines, fmt.Sprintf("Package: %s", pkg.Package))
	lines = append(lines, fmt.Sprintf("Version: %s", pkg.Version))
	lines = append(lines, fmt.Sprintf("Architecture: %s", pkg.Architecture))

	pushOpt(&lines, "Maintainer", pkg.Maintainer)
	if pkg.InstalledSize > 0 {
		lines = append(lines, fmt.Sprintf("Installed-Size: %d", pkg.InstalledSize))
	}
	pushOpt(&lines, "Depends", pkg.Depends)
	pushOpt(&lines, "Recommends", pkg.Recommends)
	pushOpt(&lines, "Suggests", pkg.Suggests)
	pushOpt(&lines, "Conflicts", pkg.Conflicts)
	pushOpt(&lines, "Breaks", pkg.Breaks)
	pushOpt(&lines, "Replaces", pkg.Replaces)
	pushOpt(&lines, "Provides", pkg.Provides)
	pushOpt(&lines, "Section", pkg.Section)
	pushOpt(&lines, "Priority", pkg.Priority)
	pushOpt(&lines, "Homepage", pkg.Homepage)

	// Extra keys need to be sorted for deterministic behavior
	var extraKeys []string
	for k := range pkg.Extra {
		extraKeys = append(extraKeys, k)
	}
	sort.Strings(extraKeys)
	for _, k := range extraKeys {
		lines = append(lines, fmt.Sprintf("%s: %s", k, pkg.Extra[k]))
	}

	lines = append(lines, fmt.Sprintf("Filename: %s", pkg.Filename))
	lines = append(lines, fmt.Sprintf("Size: %d", pkg.Size))
	lines = append(lines, fmt.Sprintf("MD5sum: %s", pkg.MD5sum))
	lines = append(lines, fmt.Sprintf("SHA1: %s", pkg.SHA1))
	lines = append(lines, fmt.Sprintf("SHA256: %s", pkg.SHA256))
	pushOpt(&lines, "Description", pkg.Description)

	return strings.Join(lines, "\n")
}

func pushOpt(lines *[]string, key, value string) {
	if strings.TrimSpace(value) != "" {
		*lines = append(*lines, fmt.Sprintf("%s: %s", key, value))
	}
}

func required(fields map[string]string, key string) (string, error) {
	val, ok := fields[key]
	if !ok || strings.TrimSpace(val) == "" {
		return "", fmt.Errorf("package entry missing required '%s' field", key)
	}
	return val, nil
}

func defaulted(value, def string) string {
	if strings.TrimSpace(value) == "" {
		return def
	}
	return value
}

func CalculateHashes(data []byte) (string, string, string) {
	md5Sum := md5.Sum(data)
	sha1Sum := sha1.Sum(data)
	sha256Sum := sha256.Sum256(data)
	return hex.EncodeToString(md5Sum[:]), hex.EncodeToString(sha1Sum[:]), hex.EncodeToString(sha256Sum[:])
}

func CompressGzip(data []byte) ([]byte, error) {
	var buf bytes.Buffer
	w := gzip.NewWriter(&buf)
	_, err := w.Write(data)
	if err != nil {
		return nil, err
	}
	w.Close()
	return buf.Bytes(), nil
}
